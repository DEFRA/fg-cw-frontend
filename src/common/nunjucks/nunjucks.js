import Vision from "@hapi/vision";
import Nunjucks from "nunjucks";

import { glob } from "node:fs/promises";
import { config } from "../config.js";
import { filters } from "../format/format-filters.js";
import { context } from "./context/context.js";
import { formatDate } from "./filters/format-date.js";

const njkPaths = [
  "node_modules/govuk-frontend/dist/",
  ...(await Array.fromAsync(
    glob(["**/views", "**/*(layouts|components|pages|partials)/"], {
      exclude: ["node_modules", "*.*"],
    }),
  )),
];

const viewPaths = await Array.fromAsync(
  glob(["**/views"], {
    exclude: ["node_modules", "*.*"],
  }),
);

export const environment = Nunjucks.configure(njkPaths, {
  autoescape: true,
  throwOnUndefined: false,
  trimBlocks: true,
  lstripBlocks: true,
  watch: config.get("nunjucks.watch"),
  noCache: config.get("nunjucks.noCache"),
  filter: {
    formatDate,
  },
});

environment.addFilter("formatDate", formatDate);
environment.addFilter("fixed", filters.fixed);

export const nunjucks = {
  plugin: Vision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = Nunjucks.compile(src, options.environment);
          return (ctx) => template.render(ctx);
        },
      },
    },
    compileOptions: {
      environment,
    },
    relativeTo: config.get("root"),
    path: viewPaths,
    isCached: config.get("isProduction"),
    context,
  },
};
