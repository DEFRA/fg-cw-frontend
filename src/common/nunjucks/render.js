import camelCase from "lodash/camelCase.js";
import { environment } from "./nunjucks.js";

/**
 * @param {string} name
 * @param {object} params
 * @param {string} [block]
 */
export const render = (name, params, block) => {
  const macro = {
    path: `${name}/macro.njk`,
    name: camelCase(name.split("-")),
    params: JSON.stringify(params, null, 2),
  };

  const callComponent = block
    ? `{%- call ${macro.name}(${macro.params}) -%}${block}{%- endcall -%}`
    : `{{- ${macro.name}(${macro.params}) -}}`;

  return environment.renderString(
    `{%- from "${macro.path}" import ${macro.name} -%}
    ${callComponent}
    `,
    {},
  );
};
