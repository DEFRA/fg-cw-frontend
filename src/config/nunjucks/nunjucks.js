import { fileURLToPath } from 'node:url'
import path from 'path'
import nunjucks from 'nunjucks'
import hapiVision from '@hapi/vision'

import { config } from '../config.js'
import { context } from './context/context.js'
import * as filters from './filters/filters.js'
import * as globals from './globals.js'

const dirname = path.dirname(fileURLToPath(import.meta.url))
const nunjucksEnvironment = nunjucks.configure(
  [
    'node_modules/govuk-frontend/dist/',
    path.resolve(dirname, '../../server/common/templates'),
    path.resolve(dirname, '../../server/common/components'),
    path.resolve(dirname, '../../case-management/views'),
    path.resolve(dirname, '../../case-management/views/layouts'),
    path.resolve(dirname, '../../case-management/views/components'),
    path.resolve(dirname, '../../case-management/views/pages'),
    path.resolve(dirname, '../../case-management/views/partials'),
    path.resolve(dirname, '../../common/presentation')
  ],
  {
    autoescape: true,
    throwOnUndefined: false,
    trimBlocks: true,
    lstripBlocks: true,
    watch: config.get('nunjucks.watch'),
    noCache: config.get('nunjucks.noCache')
  }
)

export const nunjucksConfig = {
  plugin: hapiVision,
  options: {
    engines: {
      njk: {
        compile(src, options) {
          const template = nunjucks.compile(src, options.environment)
          return (ctx) => template.render(ctx)
        }
      }
    },
    compileOptions: {
      environment: nunjucksEnvironment
    },
    relativeTo: path.resolve(dirname, '../..'),
    path: ['server', 'case-management/views', 'common/presentation'],
    isCached: config.get('isProduction'),
    context
  }
}

Object.entries(globals).forEach(([name, global]) => {
  nunjucksEnvironment.addGlobal(name, global)
})

Object.entries(filters).forEach(([name, filter]) => {
  nunjucksEnvironment.addFilter(name, filter)
})
