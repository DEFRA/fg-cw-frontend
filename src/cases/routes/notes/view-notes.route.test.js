import hapi from "@hapi/hapi";
import { load } from "cheerio";
import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { createMockCaseData } from "../../../../test/data/case-test-data.js";
import { nunjucks } from "../../../common/nunjucks/nunjucks.js";
import { findCaseByIdUseCase } from "../../use-cases/find-case-by-id.use-case.js";
import { viewNotesRoute } from "./view-notes.route.js";

vi.mock("../../use-cases/find-case-by-id.use-case.js");

describe("viewNotesRoute", () => {
  let server;

  beforeAll(async () => {
    server = hapi.server();
    server.route(viewNotesRoute);
    await server.register([nunjucks]);

    await server.initialize();
  });

  afterAll(async () => {
    await server.stop();
  });

  it("calls use case with correct caseId", async () => {
    const mockCaseData = createMockCaseData();

    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
    });

    expect(findCaseByIdUseCase).toHaveBeenCalledWith(
      "68495db5afe2d27b09b2ee47",
    );
  });

  it("renders page and returns 200 ok", async () => {
    const mockCaseData = createMockCaseData();

    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("handles case with no comments", async () => {
    const mockCaseData = createMockCaseData({
      comments: [],
    });

    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });

  it("handles case with null comments", async () => {
    const mockCaseData = createMockCaseData({
      comments: null,
    });

    findCaseByIdUseCase.mockResolvedValue(mockCaseData);

    const { statusCode, result } = await server.inject({
      method: "GET",
      url: "/cases/68495db5afe2d27b09b2ee47/notes",
    });

    expect(statusCode).toEqual(200);

    const $ = load(result);
    const view = $("#main-content").html();

    expect(view).toMatchSnapshot();
  });
});
