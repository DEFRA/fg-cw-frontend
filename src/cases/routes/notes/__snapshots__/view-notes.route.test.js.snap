// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`viewNotesRoute > handles case with no comments 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <h2 class="govuk-heading-l govuk-!-margin-top-6">Notes</h2>

    <a href="notes/new" class="govuk-button govuk-button--primary">
      Add note
    </a>

  </div>
        "
`;

exports[`viewNotesRoute > handles case with null comments 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <h2 class="govuk-heading-l govuk-!-margin-top-6">Notes</h2>

    <a href="notes/new" class="govuk-button govuk-button--primary">
      Add note
    </a>

  </div>
        "
`;

exports[`viewNotesRoute > renders page and returns 200 ok 1`] = `
"
  
<div class="govuk-!-margin-bottom-6 app-case-banner">
  <div class="govuk-width-container">
    <a href="/cases" class="govuk-back-link">Back to cases</a>

    <h1 class="govuk-heading-m">
      Valley Farm
    </h1>

    <div>
      <div class="app-case-row">
      </div>
    </div>
  </div>
</div>


  <div class="govuk-body">
    
  <div class="govuk-service-navigation" data-module="govuk-service-navigation">
      <div class="govuk-width-container">

    <div class="govuk-service-navigation__container">
      

      
        <nav aria-label="Menu" class="govuk-service-navigation__wrapper">
          <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="navigation" hidden="">
            Menu
          </button>

          <ul class="govuk-service-navigation__list" id="navigation">

            
              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47">
                                    
Tasks
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/case-details">
                                    
Case Details
                  </a>
              </li>

              
              <li class="govuk-service-navigation__item govuk-service-navigation__item--active">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/notes" aria-current="true">
                                    
                  <strong class="govuk-service-navigation__active-fallback">Notes</strong>

                  </a>
              </li>

              
              <li class="govuk-service-navigation__item">
                  <a class="govuk-service-navigation__link" href="/cases/68495db5afe2d27b09b2ee47/timeline">
                                    
Timeline
                  </a>
              </li>

            </ul>
        </nav>
    </div>

    </div>

  </div>



    <h2 class="govuk-heading-l govuk-!-margin-top-6">Notes</h2>

    <a href="notes/new" class="govuk-button govuk-button--primary">
      Add note
    </a>


      <table class="govuk-table" data-module="moj-sortable-table">
  <thead class="govuk-table__head">
    <tr class="govuk-table__row">
      <th scope="col" class="govuk-table__header" aria-sort="descending">Date</th>
      <th scope="col" class="govuk-table__header" aria-sort="ascending">Type</th>
      <th scope="col" class="govuk-table__header govuk-!-width-one-half">Note</th>
      <th scope="col" class="govuk-table__header" aria-sort="ascending">Added by</th>
    </tr>
  </thead>
  <tbody class="govuk-table__body">
    <tr class="govuk-table__row">
      <td class="govuk-table__cell" data-sort-value="2025-01-01"><a id="note-undefined"></a>01 Jan 2025</td>
      <td class="govuk-table__cell">NOTE_ADDED</td>
      <td class="govuk-table__cell">
<div data-module="expandable-text" data-truncate-length="100" data-expand-text="read full text">
    <span class="expandable-text__content">This is a test note</span>
</div>
</td>
      <td class="govuk-table__cell">John Smith</td>
    </tr>
  </tbody>
</table>

  </div>
        "
`;
